### DISCUSSION TITLE
Issue with linking a city hub

### DISCUSSION THREAD

**Discussion by TekyTeky** [Novakasa/brickrail#107] [General]
I have 2 city hubs, one on a train and one as a switch hub.
I also have a technics hub that I use for another switch.
The city hub for the train works ok and also the technic hub. 
BrickRail thinks the other city hub is a technics hub. I have tried twice 2 install the firmware, the same as for the other city hub. I also reset it to original Lego firmware and then loaded again but it just won't acknowledge that it's a city hub.
I am on holidays so don't have any other hubs to test.. normally I use technic hubs for switches, so it's no big deal.
It connects ok, when selecting Add layout controller hub.
Then when I select Control devices I get a red connection error, saying it's a technics hub, but the red error is so small and it changes too fast to get it. It then locks the gui and I can't run trains.
get any ideas?

**Comment by Novakasa** [comment]
thanks for giving this a shot!

Yeah, this is a bug, I'll fix it later today. Never tested hub-type agnostic operations.
For now, you could quite easily work around this by deleting lines 6 and 57  in `ble-server-{OS}/hub_programs/layout_controller.py` in your brickrail folder.
It's the lines
```python
from pybricks.hubs import TechnicHub
```
and
```python
self.hub = TechnicHub()
```

which should be unnecessary.

**Comment by Novakasa** [comment]
Just published the release with the fix:
https://github.com/Novakasa/brickrail/releases/tag/v1.0.0-alpha.2

Thanks for reporting!

**Reply by TekyTeky** [reply]
Thanks for the quick reply and fix!

Is there a way to make the GUI window font larger. I tried changing the pc resolution once the window is open and it holds for a short time then changes back. It makes it really difficult to use or see the words to get your mouse on menu items etc?

**Reply by Novakasa** [reply]
Could you show me a screenshot of how it looks on your system compared to other apps? It might be a dpi issue. What dpi scale do you use in your OS settings? I could try out some things but it might be difficult on linux

**Reply by Novakasa** [reply]
Also which OS are you using?

**Reply by Novakasa** [reply]
I did some testing and it should be possible to implement a text scaling setting, but it'll be a bit finnicky to implement.

**Reply by Novakasa** [reply]
I've just pushed a version with "hidpi" setting turned off (no text scaling yet), it should be online soon here: https://github.com/Novakasa/brickrail/actions/runs/4769178723
(scroll down to "artifacts", download the artifact `brickrail-<OS>`

Could you report if that already fixes your problem?

**Reply by TekyTeky** [reply]
Ok, sure, I am on WIN11.. I did some before screenshots, will check it out in a min.

**Reply by TekyTeky** [reply]
Great work man.. much better. It is now the same size as the CMD BLEServer window.

Here is a before and after pic.
![Screenshot 2023-04-22 080630](https://user-images.githubusercontent.com/87248517/233743769-7b418acc-d034-49fc-89f3-359e02a135f9.png)
![Screenshot 2023-04-22 082256](https://user-images.githubusercontent.com/87248517/233743791-df51643b-5982-4fa6-a267-4a6d55b182e8.png)

**Reply by TekyTeky** [reply]
I'll try the Hub update later, it's only 8:30am Saturday in Sydney, Aus.

**Reply by TekyTeky** [reply]
OK, the City Hub works now.. thanks!!

I have finally figured out the + and - symbols and how they work and affect the train direction.

What you were saying about the Chroma of the colors of the bricks, I am using Orange and that works ok too, but I have left the physical switch color to None. I guess that's ok? I am also using Red and Blue and have labeled them.

Actually, if I turn the train around and run it backwards, - to - then the train runs slower and sees the Orange, if I run it forward, + to + then it runs faster and misses the Orange. So, I'll just use Red or Blue.
But this is if I have the train set to Oneway.. I just don't get the + and -.
I'll re read the instructions.

**Reply by TekyTeky** [reply]
I have some questions:
In the instructions it says you can label a Switch, but that seems to have gone.
I can't get the Switches to change when running "manually", the train works fine.
How does the 'Connect Hubs' and 'Disconnect Hub's work?
How does the "Speed" next to Random targets affect the train?
Battery levels would be good so you know when to swap batteries, I use high amp rechargeables. I think Batt levels is part of PyBricks.

Lastly, for the moment ;-)
It would be good to be able to label a block. I have several rail yards at the end of lines and they are currently labeled so I know where to send them from the track plan or coal yard or station etc etc.
I have some trains with 2 motors, so they are linked... is that possible?
Also some MOCs have the motor the wrong way around, is it possible to have an "Invert" option like with the Switches?

Please don't feel obliged to add anything that isn't fixing a bug.
I am just trying to get it to a similar stage to my BAP setup which has 8 Technic Hubs and about 20 Locos, but with my current PC with 16gb of Ram I can only run 10 BLE devices at any one time. With BAP I have sensors on the track to denote 'Sections' the start and finish.. so am limited. I have extended the cables for the Switches and Sensors myself so they reach.

You have done a great job with this.. thanks so much.
Cheers
John

**Reply by Novakasa** [reply]
> What you were saying about the Chroma of the colors of the bricks, I am using Orange and that works ok too, but I have left the physical switch color to None. I guess that's ok? I am also using Red and Blue and have labeled them.

You're talking about "switch color" but I'm guessing you mean the "sensor colors".
Yeah, orange should also work. Color "None" just means the train doesn't care what color it actually is, but rather it detects brick with high chroma value. Chroma is high for high brightness and high saturation colors. E.g. red, blue, bright green. Low chroma is e.g. white, gray, black.

See this HSV-cone, high chroma means high distance to the central axis:
![image](https://user-images.githubusercontent.com/59559612/233771303-20336778-3565-4e3e-a04a-85a72463c42f.png)

If you set the color not to "None", but a specific color, the train will actually expect this specific color, and if it doesn't match, the layout will stop everything. So it is basically a safety feature if things go wrong. But, I don't think that I have orange as a standard color yet.

"None" is the default, and everything should work as expected. Make sure that you don't have other colors on your layout that don't correspond to an actual "sensor" in Brickrail. By default, you only have 2 Sensors per block.

> Actually, if I turn the train around and run it backwards, - to - then the train runs slower and sees the Orange, if I run it forward, + to + then it runs faster and misses the Orange. So, I'll just use Red or Blue. But this is if I have the train set to Oneway.. I just don't get the + and -. I'll re read the instructions.

Currently, I do recommend having the trains set to "only forward", since I have done the most testing with that. One issue with trains going backwards is that they don't have a opportunity to slow down before they need to stop if the block has no "prior sensors". See this section in the wiki for more details: https://github.com/Novakasa/brickrail/wiki/3-Create-a-virtual-layout#layout-sensors

Like I said above, setting the sensor color to "None" should work perfectly, if you set it to a different Color than you are actually using, you might get problems.

>In the instructions it says you can label a Switch, but that seems to have gone.

I can't remember writing that, where did you see that in the instructions? Do you mean Blocks?

> I can't get the Switches to change when running "manually", the train works fine.

I'm guessing you mean when "control devices" is off? Yeah, I am planning to have a mode where you can still control switches but are allowed to run the trains manually. This is tracked in #77

> How does the 'Connect Hubs' and 'Disconnect Hub's work?

On a low level, each hub can have 3 different states: "disconnected", "connected", "program running".

The last state means that the correct micropython program is running on the hub, which is required for operation.

You can manually control connecting and running the program in the left panel. But most of the time, you want to simply have every hub connected and running the program, so the "Connect Hubs" button simply connects all hubs that aren't currently connected, and runs the program on all hubs that currently don't have the program running. When you enable "control devices" this is done automatically.

"Disconnect hubs" simply does the opposite, it stops all running hubs and disconnects them. This happens automatically when you quit Brickrail.

> How does the "Speed" next to Random targets affect the train?

This makes the virtual trains go faster or slower, but doesn't actually affect the physical speed of your physical trains. It is mostly intended as a gimmick when running a layout virtually ( = "simulating" the layout). You can make the trains go extremely fast if you set the speed to e.g. 10. The only way this currently affects your physical layout is that when "random targets" is turned on, the wait time is affected by this speed setting. So if you want the trains to wait longer with "random targets", set it to a number lower than 1, e.g. 0.3.

> Battery levels would be good so you know when to swap batteries, I use high amp rechargeables. I think Batt levels is part of PyBricks.

I agree, I'll open an issue for it

> It would be good to be able to label a block. I have several rail yards at the end of lines and they are currently labeled so I know where to send them from the track plan or coal yard or station etc etc.

This actually was implemented until a few weeks ago, but I removed it because I didn't like the way the font looked on the otherwise clean layouts. I will definitely try to figure out a better solution in the future, or maybe just make the labels optional in the settings.

> I have some trains with 2 motors, so they are linked... is that possible?

This is a bit tricky, because with a city hub you only get 2 ports. And currently one of them is required to be used for the Sensor. Do some of your trains have other hubs? Then it would be principally possible to do something, but it is not high on the priority since I do not have big trains that require two motors in one loco.

In the future, it might be possible to also setup some stationary sensors instead of the sensor on the train, but that is very low priority for me, since for me I don't have many trains at the same time, but would like a complex track layout, so it is much less sensors if the sensors are on the train. Mixing both approaches is probably not possible, so you would either need a full layout controller with stationary sensors or have every train equipped with a sensor.

> Also some MOCs have the motor the wrong way around, is it possible to have an "Invert" option like with the Switches?

Yeah this is definitely planned, and hopefully soon.

> Please don't feel obliged to add anything that isn't fixing a bug. I am just trying to get it to a similar stage to my BAP setup which has 8 Technic Hubs and about 20 Locos, but with my current PC with 16gb of Ram I can only run 10 BLE devices at any one time.

Wow that's impressive, 20 locos... Would love to see a picture of your setup!

Though I am a bit doubtful that 16 GB is the reason for the limit. It might actually the bluetooth chips or the interference that limits how many hubs you can run. So the amount of hubs might be the same with Brickrail.

> You have done a great job with this.. thanks so much. Cheers John

Thanks. It's great to see people using it, and especially having a power user like you :)

**Reply by Novakasa** [reply]
Also I encourage you to play around with controlling trains in a purely virtual layout. This might help in understanding how the different logical blocks work (+/-) and how the train going backwards affects things.

Another thing that might be confusing at fist: A train will always try to position itself _in front_ of the color marker compared to its orientation. That means, if your train is arriving backwards in a block, passing the final marker, it will change direction and drive slowy forwards until it has passed the same marker again. Internally, this means that the train will always try to achieve "positive facing direction". This means that the train orientation and the last direction it went is aligned.

**Reply by TekyTeky** [reply]
Thanks for all that info. I have spent most of the day "playing" around with the layout, so I think I understand it more now.
Reversing is an issue, so I logged another discussion explaining... it may help for the future.

It was naming Blocks... not Switches... under Creating Blocks it says:
 You will notice the "create block" button in the inspector panel to the right of the viewport. Click this button, type in a name and click OK to create the block.
But that's fine.

Your last comment now makes sense with what I saw... it seems it has issues for reversing, but only sometimes.
It is knowing where a prior should be in a block. 
I plan on redoing my layout at some stage, maybe next year, to make it less congested and hopefully bigger radius curves... hence why I am interested in Brickrail.
I tried Mattzo Bricks but couldn't get my Arduinos to work.

The one thing I like about BAP is that you can make the Switches work using the app but still use the Lego remotes to control the trains. There is a function to only connect to the MAC addresses of particular Hubs.

I have a lot of the color PU Lego sensors as I was going to add them to locos to trigger sounds with color bricks in the track. Which I do sometimes.
I use the older motion sensors from WeDo2, as they are black and white so good as an end or section sensor.

Here are some pics I have taken today. It's difficult to see but the whole city is lit.. and some of the carriages too, using old PF battery packs and Led strips.
Many are the Lego Trains and Modulars, but lots are replicas of real locos or from old Lego trains I have copied.

![IMG_9246](https://user-images.githubusercontent.com/87248517/233780190-16a343c1-0adb-42bf-9eff-0e2dbad28f68.jpeg)
![IMG_4215](https://user-images.githubusercontent.com/87248517/233780151-80525918-2b20-4240-b752-bab418eb6314.jpeg)
![IMG_9240](https://user-images.githubusercontent.com/87248517/233780154-b01f40dc-3db2-46c9-b971-87840d5d06cb.jpeg)
![IMG_9241](https://user-images.githubusercontent.com/87248517/233780160-35bc5eb8-8a93-4b84-a786-ebabe02d54d7.jpeg)
![IMG_9242](https://user-images.githubusercontent.com/87248517/233780163-59042a81-c9d3-4b0c-8839-d345b10036f7.jpeg)
![IMG_9243](https://user-images.githubusercontent.com/87248517/233780169-e5cb8386-bcb8-4f3f-9e1d-f8d106e8dc75.jpeg)
![IMG_9244](https://user-images.githubusercontent.com/87248517/233780175-0c28bb56-8904-4301-8035-8275ec1a8dac.jpeg)
![IMG_9245](https://user-images.githubusercontent.com/87248517/233780184-2d8a458a-8585-4005-997f-81bcd60b2908.jpeg)

**Reply by Novakasa** [reply]
Your layout looks amazing!!

I also have in back of my mind the ability to manually control the trains with remote controls while the switches are controlled by brickrail. I need to think about that further though, since by default you will still have the pybricks firmware installed so adding remote support to that would probably make the programs too large.

**Comment by TekyTeky** [comment]
I had an odd issue, that has happened a few times. The train will connect ok but when I click Connect Devices the train looks like it connects in the CMD window but says it fails.
![Screenshot 2023-04-27 205806](https://user-images.githubusercontent.com/87248517/235002524-c5986156-14f7-44ac-b69d-913644abca85.png)

**Reply by Novakasa** [reply]
Yeah we only wait 10 seconds for the hub program to start, I will increase that since sometimes with multiple hubs and large programs the download can take quite long. (Though with the current firmware setup, I never took so long, so it probably depends on specific bluetooth hardware)

I am currently working on changing the whole download and run approach. Hopefully in the future, the download needs to happen only once per hub and firmware, and after that it just has to be started. I am also investigating whether we can drop the custom firmware and just use the regular pybricks firmware, since the resulting bigger downloads wouldn't matter so much, when we only download once.

**Reply by TekyTeky** [reply]
Ah, ok, I did a sneaky and downloaded the version with the Batt level and speed control... but the timeout fails all the time for my train.
![Screenshot 2023-04-28 165632](https://user-images.githubusercontent.com/87248517/235078405-4d7e9ff2-9d44-46ac-bfea-eff823590a44.png)

**Reply by TekyTeky** [reply]
Oh, also I need to restart the app .. disconnecting the train doesn't work .

**Reply by Novakasa** [reply]
Yeah, download on current master takes even longer currently, hopefully I can manage implementing the new download behavior soon, or I need to make users install new firmware.

**Reply by Novakasa** [reply]
I just increased the timeout in master so this new version might work better for you: (but it's still quite untested)
https://github.com/Novakasa/brickrail/actions/runs/4828083685

**Reply by TekyTeky** [reply]
Not sure if I mentioned elsewhere, but that worked.. thanks. 
Odd that if they are the same firmware that the train takes longer.
Using std Pybricks would make updates easier I guess. 

Oh I kept meaning to ask, when I download from the artefacts link there is the BLE zip file.. I download it and unzip with the main file but am I supposed to do anything with it?

**Reply by Novakasa** [reply]
no the ble_server artifact you don't need, it is just there because building the release is a 2 step process, and the result from the ble-server artifacts will be used and put in the brickrail artifact.

**Comment by TekyTeky** [comment]
With the current layout I only need one hub (Technic Hub), but I can't seem to delete the other one (City hub)
If I don't connect it then the Control Devices fails as it keeps looking for it.
Even though I have connected the other switch to the technic hub, port B... and updated the switch details.

**Reply by TekyTeky** [reply]
As I have said elsewhere, I have deleted the old layout and redoing it so the above isn't relevant atm. But going forward people would need to be able to remove a physical hub from a layout.

**Reply by Novakasa** [reply]
Yeah, definitely agree, I'll implement that soon

**Reply by Novakasa** [reply]
Implemented now in master

**Comment by TekyTeky** [comment]
I removed my green train and added a city train... but when I went to load the FW I see there are now 2... 73 and 2730.
I installed 2730, but it wouldn't connect in the control trains section, it kept timing out... so will try the other 73 version.

**Reply by TekyTeky** [reply]
Ah, I have the motor on B instead of A...

**Reply by Novakasa** [reply]
The newer firmware is actually just normal pybricks beta firmware. The trains should work with that, but the layout controller not yet.

**Reply by Novakasa** [reply]
But it only works with current master, older alpha releases won't work with this firmware.

**Reply by TekyTeky** [reply]
So which should I use 73 or 2730.

**Reply by Novakasa** [reply]
Use 73 for now
